website/src/app/blog/how-dns-based-traffic-routing-works/readme.mdx

```diff
@@ -0,0 +1,372 @@
+configuration management problem. So let's solve it!
+
+## DNS-based traffic routing
+
+By now you're probably thinking, "this is all well and good, but what does this
+have to do with Firezone?". Glad you asked!
+
+Firezone's approach to DNS was designed to address both the insecurity of DNS
+and the shortcomings of Split DNS, into a single solution that's just as secure
+as Split DNS, but much more flexible and easier to manage.
+
+### How it works
+
+When a user signs in, the Firezone Client configures its host operating system
+to use a special, lightweight DNS proxy running inside the Client for all DNS
+queries on the system.
+
+This proxy lives at `100.100.111.0/24` for IPv4 queries and
+`fd00:2021:1111:8000:100:100:111:0/120` for IPv6 queries. These addresses are
+not routable on the public internet, so there's no risk that queries for
+internal resources will leave the device unexpectedly.
+
+The IP addresses the proxy chooses to "listen on" are determined by the
+[upstream resolvers](#configuring-client-dns-upstream-resolvers) available to
+the Client. For each IPv4 upstream resolver, the proxy listens on an address in
+the `100.100.111.0/24` range, and for each IPv6 upstream resolver, the proxy
+listens on an address in the `fd00:2021:1111:8000:100:100:111:0/120` range.
+
+<p className="text-sm text-center mx-auto italic">
+  Figure 7: Firezone's embedded stub resolver
+</p>
+
+When an application makes a DNS query, the proxy checks if the name matches a
+Firezone Resource that the user has access to. If it does, the Client then
+requests a Gateway serving that Resource to resolve the DNS query. When the
+Gateway responds, the proxy will create a temporary mapping between the actual
+IP address of the Resource and the dummy IP address it returned to the
+application. From this point onward, all subsequent packets to the dummy IP
+address are forwarded to the Gateway that resolved the query. These IPs are
+persisted for the duration of the Client's session.
+
+```
+user@host:~ % dig ifconfig.net
+
+; <<>> DiG 9.10.6 <<>> ifconfig.net
+;; global options: +cmd
+;; Got answer:
+;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 51756
+;; flags: qr rd ra; QUERY: 1, ANSWER: 2, AUTHORITY: 0, ADDITIONAL: 0
+
+;; QUESTION SECTION:
+;ifconfig.net.			IN	A
+
+;; ANSWER SECTION:
+ifconfig.net.		300	IN	A	100.96.0.2
+ifconfig.net.		300	IN	A	100.96.0.1
+
+;; Query time: 262 msec
+;; SERVER: fd00:2021:1111:8000:100:100:111:0#53(fd00:2021:1111:8000:100:100:111:0)
+;; WHEN: Sat Feb 17 14:04:40 PST 2024
+;; MSG SIZE  rcvd: 86
+```
+
+<p className="text-sm text-center mx-auto italic">
+  Figure 8: An example dig command with a DNS-based Resource
+</p>
+
+If the proxy sees a query for a name that **is not** a Firezone Resource it will
+forward the query to the host's default stub resolver, as if Firezone never
+existed. This means that Clients are automatically configured for Split DNS in
```

Comment:
Not really, the host's stub resolver is not used, it's the DNS's server that is used.

Reply:
Also it might be interesting telling how we resolved the timeouts and multiple DNS servers

Reply:
Yeah those will be good to mention too, probably with a graphical aid